### PR TITLE
Google JSON Support

### PR DESCRIPTION
**Pull request by trieloff**
Please ensure your pull request adheres to the following guidelines:
- [ ] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues

Thanks for contributing!

### CODE REVIEW

File: test/google.test.js [hunk 1]
```diff
@@ -69,3 +71,145 @@ describe('Google Integration Tests', () => {
...
   }).timeout(5000);
 });
+
+function scramble(server) {
+  server.any().on('beforePersist', (_, recording) => {
+    recording.request.headers = recording.request.headers.filter(({ name }) => name !== 'authorization');
+    delete recording.request.postData;
+
+    if (recording.request.url === 'https://oauth2.googleapis.com/token') {
+      const val = JSON.parse(z.gunzipSync(Buffer.from(JSON.parse(recording.response.content.text).join(''), 'hex')));
```
Comment (tripodsan) [inline test/google.test.js:81]: shouldn't you check if the transfer-encoding is really gzip ?

File: src/google.js [hunk 2]
```diff
@@ -10,13 +10,147 @@
...
+  return null;
+}
+
+const getIdFromPath = cache(getUncachedIdFromPath);
+
+async function handleJSON(opts) {
+  const {
+    mp, log, options,
+  } = opts;
+
```
Comment (tripodsan) [inline src/google.js:94]: check if the google secrets are really present. otherwise return a 403 or 500.

File: src/google.js [hunk 3]
```diff
@@ -10,13 +10,147 @@
...
+  oAuth2Client.on('tokens', (tokens) => {
+    tokenCache = tokens;
+  });
+
+  return oAuth2Client;
+}
+
+const createCachedOAuthClient = cache(createOAuthClient);
+
+async function getUncachedIdFromPath(path, parentId, log, options) {
```
Comment (tripodsan) [inline src/google.js:44]: why don't you move this resolution to data-embed, then you don't need any google api access here....

File: src/google.js [hunk 4]
```diff
@@ -10,13 +10,147 @@
...
+      statusCode: 404,
+      body: 'spreadsheet not found',
+      headers: {
+        'cache-control': 'max-age=60',
+      },
+    };
+  }
+
+  const sheetURL = `https://docs.google.com/spreadsheets/d/${sheetId}/edit`;
+  const url = `https://adobeioruntime.net/api/v1/web/helix/helix-services/data-embed@v1/${sheetURL}`;
```
[the remark this answers is not in the document]
Comment (trieloff) [reply src/google.js:108]: We do't have to (I might do it for Excel, which has uglier URLs)

File: src/google.js [hunk 5]
```diff
@@ -73,4 +207,4 @@ function canhandle(mp) {
   return mp && mp.type === 'google';
 }
 
-module.exports = { canhandle, handle };
+module.exports = { canhandle, handle, handleJSON };
```
[the remark this answers is not in the document]
Comment (trieloff) [reply src/google.js:78]: done.

File: src/google.js [hunk 6]
```diff
@@ -10,13 +10,147 @@
...
+      headers: {
+        'cache-control': 'max-age=60',
+      },
+    };
+  }
+
+  const sheetURL = `https://docs.google.com/spreadsheets/d/${sheetId}/edit`;
+  const url = `https://adobeioruntime.net/api/v1/web/helix/helix-services/data-embed@v1/${sheetURL}`;
+
+  const response = await fetch(url, options);
```
[the remark this answers is not in the document]
Comment (trieloff) [reply src/google.js]: fixed.

File: src/google.js (hunk 2 above)
Comment (trieloff) [reply src/google.js]: The Google client checks it already and throws, which leads to a 500

**Comment by tripodsan** [comment]
> objections addressed

no, they weren't

File: src/google.js (hunk 3 above)
Comment (trieloff) [reply src/google.js]: It would require changing the data-embed API from `url` to `rootid, type, path` and add complexity there that is better kept here.

**Pull request merged**